### ISSUE TITLE
Identify no side-effect statements

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#221]
_It should be a very ambicious inspection, but I think that will be one of more powerful and helpful. But, it should be disabled by default while in development, because it could trigger a lot of false-positives._

**Consider the next code:**

```php
class Sum {
    private $numbers = [];

    public function push($number) { $this->numbers[] = $number; }
    public function calculate()   { return array_sum($this->numbers); }
    public function print()       { print($this->calculate()); }
}
```

**And the usage:**

```php
// First reference:
$sum = new Sum;
$sum->push(1);
$sum->push(2);

// Second reference:
echo $sum->calculate(); // => 3, or
$sum->print();          // => 3 too
```

Based on original class you should note that if we just write the _first reference_ code it will give any sense to our code, because will not have any effect (_no side-effect code_). It only will have an existence sense only when we write the _second reference_ code (the _side-effect code_).

In case where you don't need to write the _second reference_ code, you can just drop the _first reference_ code, because you code doesn't nothing important.

## Types of side-effects

First we should understand the types of side-effects on a method/function/statement. PHP functions and classes methods too should be mapped to make it possible (_the boring part that I will explain on next topics_). 

Each method/function/statement should receives a kind of an _invisible annotation_, as if we do something like `@side-effect {type}` annotation in each one of them - but in this case, EA will do that for us.

Now lets talk about five kinds of side-effects that I thought. Just to make reading short, where you read "method", I want to mean "a method, function or some statements like echo".

**Side-effect undefined:** (`SideEffect::UNDEFINED`)

It'll be the first state of a method. Basically is useful to the EA identify that this method still not was checked _if have or not a side-effect_.

**Side-effect unknow:** (`SideEffect::UNKNOW`)

In case where EA could not detects if a method _is side-effect or not_, then it should be marked as _unknow side-effect_. Any other method that uses an unknow side-effect method will be treated too as an unknow side-effect method.

Then, imagine that you have not mapped the function `array_sum()` internally as a _side-effect none_ (_that is explained on next topic_), then it will be treated automatically as an _unknow side-effect_ function. If you do use this function in some method, then your method too will be treated as a _unknow side-effect_ (_but I will explain the precedence schema after_).

**Side-effect none:** (`SideEffect::NONE`)

If your method does nothing to the own instance (as just read some properties and static arguments), then this is a _no side-effect method_. It mean that doesn't have any sense of you use that as a single statement call, but only if you use inside a _side-effect statement_.

For instance, the `Sum::calculate()` method (_from our example code_) is a _no side-effect_ because it just return a value. And yes, it uses a PHP function, in case, `array_sum()` that too is a _no side-effect_. Then if you just call it like: `$sum->calculate()` it'll not do nothing useful to your code. This line could be dropped. 

To fix that you should use a _side-effect function_ like `print($sum->calculate())`.

You can turn a no side-effect call into a side-effect when you stores the returned value into a variable (but it'll be a local side-effect). For instance: `$value = $sum->calculate()` is valid. Or then, if you use it directly to an argument of another method, like `print()` as I mentioned before.

**Side-effect soft:** (`SideEffect::SOFT`)

If your method modifies some internal property, change some referenced argument (`&$argument`) or globals (like `$_POST`), then your method should be treated as a _soft side-effect_. It mean that, different of the previous type, make sense you call this method without use the returned value, as a singular call.

For instance, the `Sum::push()` is a _soft side-effect_ because it will affects an internal property of instance. Then make sense you just call like `$sum->push(1)` without assign to a variable (_something that the previous type do not allow_). But the variable `$sum`, _at this specific point_, still have not sense to exists, because you do any operation with a `SideEffect::HARD` (_next topic_).

**Side-effect hard:** (`SideEffect::HARD`)

If your method outputs some kind of data _virtual or phisically_, then your method is a _hard side-effect_. It happen with functions like `fwrite()`, `print()` or statements like `echo`.

A _no side-effect_ method could be "turned into" a _hard side-effect_ when is attached to a _hard side-effect_. As mentined, the `print($sum->calculate())` make the _no side-effect_ `Sum::calculate()` to be valid because `print()` is a _hard side-effect_ function.

Then, for instance, if you method uses some _hard side-effect_ method, as `Sum::print()` does, then you could call it by itself directly. In this case, it'll make the `$sum` variable make sense to exists and will disable the inspection for it.

## Side-effect precedence

Now that I have explained all types of side-effects, we should understand that each method should have a single type: or it is _unknow_, _none_, _soft_ or _hard side-effect_ (remembering, the _undefined side-effect_ refer to an unanalyzed method, which is a _temporary type_). 

In this case, we should too understand the precedence of types. It is, if your method uses a lot of differents types of side-effects in it structure, it should be only a kind of them (_the highest_).

> **In this case, it's the precedence:** 
(Undefined →) None → Unknow → Soft → Hard

Now let me explain that with an example code:

```php
public function print () {
    // First reference:
    if (!$this->numbers) {
        return null;
    }

    // Second reference:
    $sum = array_sum($this->numbers);

    // Third reference:
    $this->sum = $sum;

    // Fourth reference:
    print($this->sum);
}
```

Okay. Let's supposes that we are writing our code in steps (from _first reference_ to _fourth reference_). While we are coding the _first reference_ code, our method `print()` is treated like a `SideEffect::NONE` because it just make some sense if you do something like `echo $sum->print()`.

Now we write up the _second reference_ code (_1 and 2_), maintaining the assumption that `array_sum()` was not mapped on EA as a `SideEffect::NONE` (and therefore, is a `SideEffect::UNKNOW`), then, because of the precedence of that `SideEffect::UNKNOW` is greatest than `SideEffect::NONE`, our method at this point is too a `SideEffect::UNKNOW` (_which makes the side-effect analysis impossible on `$sum` instance_).

Now we write up the _third reference_ code (_1, 2 and 3_), as we are modifing an internal property, then at this point our method is marked as a `SideEffect::SOFT` (_now the side-effect analysis on `$sum` instance is possible again_).

Finally, after write up all references and completes our method structure, we finally specify a method that is a `SideEffect::HARD` (because it uses `print()` that is a `SideEffect::HARD`). This is the highest level of side-effect and cannot be overrided by other types (if I declare a `SideEffect::SOFT` after, for instance, it still should be treated as a `SideEffect::HARD`).

Well, except if you consider a new annotation type like `@side-effect {type}` that will forces EA to use this specification as the method side-effect type. Then if you declares it as `@side-effect soft`, the EA analysis will not be done in this method and will just consider what you have annotated.

## Conditional side-effect

There some cases where we could have some _conditional side-effect_. In this case, we should consider that conditional is always true (even for _else_ or _elseif_) and capture the highest precedence on it. 

Then, for instance, if we have:

```php
public function print (bool $print = null) {
    if ($print === true) {
        print($this->message);
    }
}
```

We should consider that this method is a `SideEffect::HARD` because it uses the `print()` that too is (even that the argument is not _true_). _It could be fixed on future on a next visit to this inspection, because I think that will be very hard to make it work._

## PHP function, methods and statements mapping

A very important thing is map all PHP functions, methods (_from internal classes, of course_) and ouput statements like echo (_can't imagine other, for now_). It will be the most boring step, because PHP have a lot of functions and each one should be mapped, else it should be treated as `SideEffect::UNKNOW`. I could help on this step by check each PHP functions and work on the file that specifies that (_suggestion below_).

For now, we should always consider the highest precedence of an internal method, when it could have two differents states because of type of parameter. Currently I have no function like that on my mind, but imagine that we have a PHP function that is `something($a, $b = false)` that is `SideEffect::NONE` when `$b = false` and `SideEffect::HARD` when `$b = true`. In this case, we should consider this function as `SideEffect::HARD` (that is, _respecting the precedence_). Futurely we could check each exceptions like that _one-by-one_, and determine the right side-effect based on number or type of arguments.

My suggestion for this mapping is create a `HashMap` on this inspection class that receives a list the PHP internal functions and methods, then we could implements that easily. Something like that (_beware, pseudo-code_):

```java
private static final HashMap<String, SideEffect> internalsMapping;

static
{
    // Functions.
    internalsMapping = new HashMap<String, SideEffect>();
    internalsMapping.put("array_sum", SideEffect::NONE);
    internalsMapping.put("array_push", SideEffect::SOFT);

    // Classes...
    internalsMapping.put("ArrayObject::offsetGet", SideEffect::NONE);
}
```

Remembering, any method outside this list that is not a user code should be treated as `SideEffect::UNKNOW`. Structures like `echo` should be managed directly by inspection.

## About vendor methods

In case where you have vendor methods, you should parse each used methods and functions normally, as an user code (it's is too, _at vendor context_). Then EA should check the vendor method and mark it side-effect type when it need to be done.

## Optimization

You could imagine that it will be a _very very very heavy process_, but I think that is not, because we can apply some kind of optimizations. The most important is determine the side-effect type once or after method be updated - then should trigger the inspection in each place that you use that method (_what happen naturally on PS, I guess_).

After that, the inspection should just check how the code was wrote to determine if is a side-effect then report or not that. Then, for instance:

```php
$sum = new Sum;

// Sum::__construct() is SideEffect::UNDEFINED, analyzing... (done)
// Sum::__construct() doesn't exists, then is SideEffect::NONE.

// At this point, $sum could be dropped, because Sum have no __construct (or even if the __construct is a SideEffect::SOFT or lower), then it is marked as a SideEffect::NONE.

$sum->push(1);

// Sum::push() is SideEffect::UNDEFINED, analyzing... (done)
// Sum::push() is SideEffect::SOFT, because just update an internal property.

// At this point, $sum still can be dropped, because both lines is SideEffect::NONE.

$sum->calculate();

// Sum::calculate() is SideEffect::UNDEFINED, analyzing... (done)
// Sum::calculate() is SideEffect::SOFT, because it return an internal property.

// At this point, the $sum variable was marked as a SideEffect::SOFT, what is insufficient to be exists. Then all code can be dropped securely.

echo $sum->calculate();

// Sum::calculate() is SideEffect::SOFT (identified in previous analysis)
// Sum::calculate()->parent is instance of EchoStatement, that is a SideEffect::HARD (mapped internally).

// At this point, the $sum variable was remarked to be a SideEffect::HARD, then now the code is valid and have a real effect.
```

---

_Uff! I just done, I guess... As I understand that it is a complex implementation (If we can consider that its implementation is feasible), then I will be happy to discuss that with you._

**Comment by kalessil** [comment]
`$sum->calculate()` can be also spotted by #72. Of course, it's not as a detailed concept as here, but still an efficient shortcut (enabled by default, 100% accuracy).

Speaking honestly, full implementation is not feasible due to the OSS nature of the project.

**Comment by rentalhost** [comment]
@kalessil Exactly. I thought I had sent this issue to the PS team (the #72), then maybe you can consider the previous closed, because this one is more detailed.

I think that this project can accept thing like that because it's a step of inspection. **It seems to be very complex** (_was what I thought initially_), but I guess that it's only _long described_.

I guess that you can implements that step-by-step. I think that the major issue is about "stores" the type of side-effect of each method, but if PS don't supports things like that, maybe you can use a `HashMap` on inspection that map each user-land methods (_like is supposes to do with PHP internals_).

**Something like** (_beware, incomplete pseudo-code_):

```java
private static final HashMap<String, SideEffect> internalsSideEffect = new HashMap<String, SideEffect>();
private static final HashMap<String, SideEffect> userSideEffects = new HashMap<String, SideEffect>();

private static SideEffect identifySideEffect (Function function) {
	final String functionSQN = function.fullSQN();

	// Check if function is internally determined.
	if (internalsSideEffect.contains(functionSQN)) {
		return internalsSideEffect.get(functionSQN);
	}
	// Check if function was modified.
	else if (function.wasModified()) {
		userSideEffects.remove(functionSQN);
	}
	// Check if function was qualified.
	else if (userSideEffects.contains(functionSQN)) {
		return userSideEffects.get(functionSQN);
	}

	// 1. Identify the function SideEffect.
	// 2. Store.
	// 3. Return.

	// Note: this method could be executed sometimes, because
	// is possible that you should identify others functions here
	// (dependecies).
}

// For function call you just check for SideEffect::NONE.
private void visitPhpFunctionCall(FunctionReference reference) {
	final SideEffect referenceSideEffect = this.identifySideEffect(reference.getFunction());

	// If side-effect is NONE, it could not be a single statement.
	// Internals note: to fixes that, uses the returned value with return or print that, for instance.
	if (referenceSideEffect == SideEffect::NONE &&
	    reference.isSingleStatement()) {
	    registerProblem();
	}
}

private void registerProblem() {
	holder.registerProblem(reference, 'You can remove this function call because it have no side-effect.', ProblemHighlightType.WARNING);	
}
```

The hardest part of this code is identify the variable current side-effect based on call position. For instance:

```php
$sum = new Sum; // $sum is SideEffect::NONE, can be dropped
$sum->push(1); // $sum is now SideEffect::SOFT, can be dropped
$sum->print(); // $sum is now SideEffect::HARD, should be keeped
```

I guess that it should be done when you are treating the method structure, but I don't know how it should be done in global scope (outside of a method or function). The logic is the same, but I don't know how works the "visitor" in this case.

Ah... the `SideEffect::UNDEFINED` seems not have sense, because it is just the absense of information (it's, it doesn't was defined on HashMap, then it should be processed yet, that is the same effect of this "type").

**Comment by rentalhost** [comment]
@kalessil Did you think that this inspection is compatible with the project? I am think to work on it (no promises, I should read a lot still).

**Comment by rentalhost** [comment]
Removing the milestone. Too complex for now. I will back to working on it next month.
